gcc/config/riscv/corev.md

```diff
@@ -31,28 +31,30 @@
 
 (define_insn "cv_mac"
   [(set (match_operand:SI 0 "register_operand" "=r")
-    (plus:SI (mult:SI (match_operand:SI 2 "register_operand" "r")
-		      (match_operand:SI 3 "register_operand" "r")
-	     )
-	     (match_operand:SI 1 "register_operand" "0")
+    (plus:SI 
+      (mult:SI (match_operand:SI 1 "register_operand" "r")
+		           (match_operand:SI 2 "register_operand" "r")
+	    )
+	    (match_dup 0)
     )
   )]
   "TARGET_COREV_MAC"
-  "cv.mac\t%0,%2,%3"
+  "cv.mac\t%0,%1,%2"
```

Comment:
@pietraferreira When it is a set, you cant use match_dup.
e.g. a = a + b * c (Here we CAN'T use mach_dup)
       b = a + a * c (Here we CAN use match_dup)
Where we can't use match_dup (i.e. where the operands that are the same are both destination and source) we can use a "0", "1" etc. in the constraints fields.

Reply:
`0', `1', `2', ... `9'
An operand that matches the specified operand number is allowed. If a digit is used together with letters within the same alternative, the digit should come last.
This number is allowed to be more than a single digit. If multiple digits are encountered consecutively, they are interpreted as a single decimal integer. There is scant chance for ambiguity, since to-date it has never been desirable that `10' be interpreted as matching either operand 1 or operand 0. Should this be desired, one can use multiple alternatives instead.

This is called a matching constraint and what it really means is that the assembler has only a single operand that fills two roles considered separate in the RTL insn. For example, an add insn has two input operands and one output operand in the RTL, but on most CISC machines an add instruction really has only two operands, one of them an input-output operand:

          addl #35,r12
     
Matching constraints are used in these circumstances. More precisely, the two operands that match must include one input-only operand and one output-only operand. Moreover, the digit must be a smaller number than the number of the operand that uses it in the constraint.

For operands to match in a particular case usually means that they are identical-looking RTL expressions. But in a few special cases specific kinds of dissimilarity are allowed. For example, *x as an input operand will match *x++ as an output operand. For proper results in such cases, the output template should always use the output-operand's number when printing the operand.

Reply:
(match_dup n)
This expression is also a placeholder for operand number n. It is used when the operand needs to appear more than once in the insn.
In construction, match_dup acts just like match_operand: the operand is substituted into the insn being constructed. But in matching, match_dup behaves differently. It assumes that operand number n has already been determined by a match_operand appearing earlier in the recognition template, and it matches only an identical-looking expression.

Note that match_dup should not be used to tell the compiler that a particular register is being used for two operands (example: add that adds one register to another; the second register is both an input operand and the output operand). Use a matching constraint (see Simple Constraints) for those. match_dup is for the cases where one operand is used in two places in the template, such as an instruction that computes both a quotient and a remainder, where the opcode takes two input operands but the RTL template has to refer to each of those twice; once for the quotient pattern and once for the remainder pattern.

Reply:
If you look at the Note in match_dup it says this

Reply:
Sorry, I notice you have done this a few times

Reply:
Thanks for reviewing, it makes sense. I thought because the specification is `rd = rd + rs1 * rs2` I could match_dup the input, but as you pointed out I can't have the input operand being both input and output. I'll fix it accordingly :)